File: sources/MockTaskNotifierTask.m
//
//  MockTaskNotifierTask.m
//  iTerm2SharedARC
//
//  Test-only mock implementation of iTermTask protocol for testing TaskNotifier behavior.
//

#import "MockTaskNotifierTask.h"
#import <fcntl.h>
#import <unistd.h>

@implementation MockTaskNotifierTask {
    NSInteger _processReadCallCount;
    NSInteger _processWriteCallCount;
    NSInteger _brokenPipeCallCount;
    NSInteger _didRegisterCallCount;
    NSInteger _writeTaskCoprocessCallCount;
    NSData *_lastCoprocessData;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _fd = -1;
        _pid = 0;
        _pidToWaitOn = 0;
        _hasCoprocess = NO;
        _coprocess = nil;
        _wantsRead = YES;
        _wantsWrite = NO;
        _writeBufferHasRoom = YES;
        _hasBrokenPipe = NO;
        _sshIntegrationActive = NO;
        _dispatchSourceEnabled = NO;
        _simulateLegacyTask = NO;
    }
    return self;
}

#pragma mark - iTermTask Required Methods

- (void)processRead {
    _processReadCallCount++;
}

- (void)processWrite {
    _processWriteCallCount++;
}

- (void)brokenPipe {
    _brokenPipeCallCount++;
}

- (void)writeTask:(NSData *)data coprocess:(BOOL)isCoprocess {
    if (isCoprocess) {
        _writeTaskCoprocessCallCount++;
        _lastCoprocessData = [data copy];
    }
}

- (void)didRegister {
    _didRegisterCallCount++;
}

#pragma mark - iTermTask Optional Methods

- (BOOL)useDispatchSource {
    return self.dispatchSourceEnabled;
}

#pragma mark - Override respondsToSelector for Legacy Simulation

- (BOOL)respondsToSelector:(SEL)aSelector {
    // If simulating a legacy task, pretend we don't implement useDispatchSource
    if (self.simulateLegacyTask && aSelector == @selector(useDispatchSource)) {
        return NO;
    }
    return [super respondsToSelector:aSelector];
}

#pragma mark - Call Count Accessors

- (NSInteger)processReadCallCount {
    return _processReadCallCount;
}

- (NSInteger)processWriteCallCount {
    return _processWriteCallCount;
}

- (NSInteger)brokenPipeCallCount {
    return _brokenPipeCallCount;
}

- (NSInteger)didRegisterCallCount {
    return _didRegisterCallCount;
}

- (NSInteger)writeTaskCoprocessCallCount {
    return _writeTaskCoprocessCallCount;
}

- (NSData *)lastCoprocessData {
    return _lastCoprocessData;
}

#pragma mark - Test Helpers

- (void)reset {
    _processReadCallCount = 0;
    _processWriteCallCount = 0;
    _brokenPipeCallCount = 0;
    _didRegisterCallCount = 0;
    _writeTaskCoprocessCallCount = 0;
    _lastCoprocessData = nil;
    _dispatchSourceEnabled = NO;
    _simulateLegacyTask = NO;
    _wantsRead = YES;
    _wantsWrite = NO;
    _hasCoprocess = NO;
    _hasBrokenPipe = NO;
}

- (BOOL)waitForProcessReadCalls:(NSInteger)count timeout:(NSTimeInterval)timeout {
    NSDate *deadline = [NSDate dateWithTimeIntervalSinceNow:timeout];
    while (_processReadCallCount < count && [[NSDate date] compare:deadline] == NSOrderedAscending) {
        [NSThread sleepForTimeInterval:0.01];
    }
    return _processReadCallCount >= count;
}

- (BOOL)waitForCoprocessWriteCalls:(NSInteger)count timeout:(NSTimeInterval)timeout {
    NSDate *deadline = [NSDate dateWithTimeIntervalSinceNow:timeout];
    while (_writeTaskCoprocessCallCount < count && [[NSDate date] compare:deadline] == NSOrderedAscending) {
        [NSThread sleepForTimeInterval:0.01];
    }
    return _writeTaskCoprocessCallCount >= count;
}

- (void)closeFd {
    if (self.fd >= 0) {
        close(self.fd);
        self.fd = -1;
    }
}

#pragma mark - Factory Methods

+ (MockTaskNotifierTask *)createPipeTaskWithWriteFd:(int *)writeFd {
    int fds[2];
    if (pipe(fds) != 0) {
        return nil;
    }

    // Set non-blocking on read end
    int flags = fcntl(fds[0], F_GETFL);
    fcntl(fds[0], F_SETFL, flags | O_NONBLOCK);

    MockTaskNotifierTask *task = [[MockTaskNotifierTask alloc] init];
    task.fd = fds[0];  // Read end

    if (writeFd) {
        *writeFd = fds[1];  // Write end
    } else {
        close(fds[1]);  // Close write end if not needed
    }

    return task;
}

@end

File: ModernTests/FairnessScheduler/Mocks/MockTaskNotifierTask.swift
//
//  MockTaskNotifierTask.swift
//  ModernTests
//
//  Test-only mock implementation of iTermTask protocol for testing TaskNotifier behavior.
//  Configurable to test both dispatch source and legacy select() paths.
//

import Foundation
@testable import iTerm2SharedARC

@objc class MockTaskNotifierTask: NSObject, iTermTask {

    // MARK: - iTermTask Required Properties

    @objc var fd: Int32 = -1
    @objc var pid: pid_t = 0
    @objc var pidToWaitOn: pid_t = 0
    @objc var hasCoprocess: Bool = false
    @objc var coprocess: Coprocess?
    @objc var wantsRead: Bool = true
    @objc var wantsWrite: Bool = false
    @objc var writeBufferHasRoom: Bool = true
    @objc var hasBrokenPipe: Bool = false
    @objc var sshIntegrationActive: Bool = false

    // MARK: - Configuration for Testing

    /// Set to true to make useDispatchSource return YES.
    /// Default is NO (use select() path).
    var dispatchSourceEnabled: Bool = false

    /// If true, this mock does NOT respond to useDispatchSource selector,
    /// simulating a legacy task that relies on select().
    var simulateLegacyTask: Bool = false

    // MARK: - Call Tracking

    private(set) var processReadCallCount: Int = 0
    private(set) var processWriteCallCount: Int = 0
    private(set) var brokenPipeCallCount: Int = 0
    private(set) var didRegisterCallCount: Int = 0
    private(set) var writeTaskCoprocessCallCount: Int = 0
    private(set) var lastCoprocessData: Data?

    // MARK: - iTermTask Required Methods

    @objc func processRead() {
        processReadCallCount += 1
    }

    @objc func processWrite() {
        processWriteCallCount += 1
    }

    @objc func brokenPipe() {
        brokenPipeCallCount += 1
    }

    @objc func write(_ data: Data!, coprocess isCoprocess: Bool) {
        if isCoprocess {
            writeTaskCoprocessCallCount += 1
            lastCoprocessData = data
        }
    }

    @objc func didRegister() {
        didRegisterCallCount += 1
    }

    // MARK: - iTermTask Optional Methods

    @objc func useDispatchSource() -> Bool {
        return dispatchSourceEnabled
    }

    // MARK: - Override respondsToSelector for Legacy Simulation

    override func responds(to aSelector: Selector!) -> Bool {
        if simulateLegacyTask && aSelector == #selector(useDispatchSource) {
            return false
        }
        return super.responds(to: aSelector)
    }

    // MARK: - Test Helpers

    func reset() {
        processReadCallCount = 0
        processWriteCallCount = 0
        brokenPipeCallCount = 0
        didRegisterCallCount = 0
        writeTaskCoprocessCallCount = 0
        lastCoprocessData = nil
        dispatchSourceEnabled = false
        simulateLegacyTask = false
        wantsRead = true
        wantsWrite = false
        hasCoprocess = false
        hasBrokenPipe = false
    }

    func wait(forProcessReadCalls count: Int, timeout: TimeInterval) -> Bool {
        let deadline = Date(timeIntervalSinceNow: timeout)
        while processReadCallCount < count && Date() < deadline {
            Thread.sleep(forTimeInterval: 0.01)
        }
        return processReadCallCount >= count
    }

    func wait(forCoprocessWriteCalls count: Int, timeout: TimeInterval) -> Bool {
        let deadline = Date(timeIntervalSinceNow: timeout)
        while writeTaskCoprocessCallCount < count && Date() < deadline {
            Thread.sleep(forTimeInterval: 0.01)
        }
        return writeTaskCoprocessCallCount >= count
    }

    func closeFd() {
        if fd >= 0 {
            close(fd)
            fd = -1
        }
    }

    // MARK: - Factory Methods

    /// Creates a pipe task with read FD assigned to the task.
    /// Returns a tuple with the task and the write FD for testing.
    /// Caller is responsible for closing both FDs.
    static func createPipeTask() -> (task: MockTaskNotifierTask, writeFd: Int32)? {
        var fds: [Int32] = [0, 0]
        guard pipe(&fds) == 0 else { return nil }

        // Set non-blocking on read end
        let flags = fcntl(fds[0], F_GETFL)
        fcntl(fds[0], F_SETFL, flags | O_NONBLOCK)

        let task = MockTaskNotifierTask()
        task.fd = fds[0]  // Read end

        return (task, fds[1])  // Write end
    }
}
